File: app/components/folio/console/tiptap/overlay/form_component.rb
# frozen_string_literal: true

class Folio::Console::Tiptap::Overlay::FormComponent < Folio::Console::ApplicationComponent
  def initialize(node:)
    @node = node
  end

  private
    def data
      stimulus_controller("f-c-tiptap-overlay-form",
                          values: {
                            autoclick_cover: should_autoclick_cover?,
                          })
    end

    def should_autoclick_cover?
      @node.class.tiptap_config[:autoclick_cover] == true
    end

    def simple_form_as
      "tiptap_node_attrs[data]"
    end

    def form(&block)
      opts = {
        url: "#form",
        method: :post,
        as: simple_form_as,
        html: {
          class: "f-c-tiptap-overlay-form__form",
          data: stimulus_action(submit: "onFormSubmit")
        },
      }

      helpers.simple_form_for(@node, opts, &block)
    end

    def buttons_model
      [
        {
          variant: :primary,
          type: :submit,
          label: t(".submit"),
        },
        {
          icon: :close,
          variant: :icon,
          data: stimulus_action(click: "onCancelClick"),
        },
      ]
    end
end

File: app/components/folio/console/tiptap/overlay/form_component.js
window.Folio.Stimulus.register('f-c-tiptap-overlay-form', class extends window.Stimulus.Controller {
  static targets = ['form']

  static values = {
    autoclickCover: Boolean
  }

  connect () {
    window.setTimeout(() => {
      const firstInput = this.element.querySelector('.form-control:not([hidden]):not([disabled])')

      if (firstInput) {
        firstInput.focus()
      }
    }, 200) // wait for css transition

    this.startReactNodes()

    if (this.autoclickCoverValue) {
      window.setTimeout(() => {
        const btn = this.element.querySelector('.f-c-tiptap-overlay-form__react-file-picker--cover .f-c-files-picker[data-f-c-files-picker-state-value="empty"] .f-c-files-picker__btn')
        if (btn) btn.click()
      }, 0)
    }
  }

  disconnect () {
    this.stopReactNodes()
  }

  startReactNodes () {
    for (const reactNode of this.element.querySelectorAll('.folio-react-wrap')) {
      window.FolioConsole.React.init(reactNode)
    }
  }

  stopReactNodes () {
    for (const reactNode of this.element.querySelectorAll('.folio-react-wrap')) {
      window.FolioConsole.React.destroy(reactNode)
    }
  }

  onFormSubmit (e) {
    e.preventDefault()
    const data = window.Folio.formToHash(e.target)

    this.dispatch('submit', { detail: { data } })
  }

  onCancelClick () {
    this.dispatch('close')
  }
})
